lib/AST/DiagnosticEngine.cpp

```diff
@@ -856,3 +865,13 @@ DiagnosticSuppression::~DiagnosticSuppression() {
   for (auto consumer : consumers)
     diags.addConsumer(*consumer);
 }
+
+BufferIndirectlyCausingDiagnosticRAII::BufferIndirectlyCausingDiagnosticRAII(
+    const SourceFile &SF)
+    : Diags(SF.getASTContext().Diags) {
+  auto id = SF.getBufferID();
+  if (!id)
+    return;
```

Comment:
Nitpick: This won't come up in practice because you *only* modify the setting using RAII, but otherwise: should it return? Or should it reset the buffer instead?

Reply:
Yes, it was reset in an earlier draft. Hmm... This question feels like it illuminates a flaw: We don't account for nested RAII's. Rather than implement a push-down stack for the buffer, let me try just outlawing such nesting for today with an assertion. That will at least detect the occurrence of such a situation.